narwhal/consensus/src/consensus.rs

```diff
@@ -133,25 +132,35 @@ impl ConsensusState {
     }
 
     #[allow(clippy::result_unit_err)]
-    pub fn try_insert(&mut self, certificate: Certificate) -> Result<(), ()> {
-        let last_committed = self
-            .last_committed
+    pub fn try_insert(&mut self, certificate: &Certificate) -> Result<(), ()> {
+        Self::try_insert_in_dag(&mut self.dag, &self.last_committed, certificate)
+    }
+
+    #[allow(clippy::result_unit_err)]
+    fn try_insert_in_dag(
+        dag: &mut Dag,
+        last_committed: &HashMap<PublicKey, Round>,
+        certificate: &Certificate,
+    ) -> Result<(), ()> {
+        let last_committed_round = last_committed
             .get(&certificate.origin())
             .cloned()
             .unwrap_or_default();
-        if certificate.round() < last_committed {
+        if certificate.round() < last_committed_round {
```

Comment:
I see your point and definitely it makes sense. This code here follows the same logic that cleans up entries after commit https://github.com/MystenLabs/sui/blob/367da1318d4761439f857c0f0ead4d498edfe963/narwhal/consensus/src/consensus.rs#L181 

where practically the last committed certificate doesn't get purged immediately . 

I would like to revisit that but maybe as follow up work - would hesitate to change this only here.

> probably we should not process it again since it is already in the dag?

yes , when we recover though the certificate doesn't already exist in the dag. But as I said above we follow the same approach we do on other parts of the code.